### DISCUSSION TITLE
Where can I see the log when adding json-prune?

### DISCUSSION THREAD

**Discussion by dimisa-RUAdList** [uBlockOrigin/uBlock-issues#3096] [Q&A] [answered]
uBlock Origin 1.55.1b21

Where can I see the log when adding json-prune?

Example: `3dnews.ru##+js(json-prune)`

**Comment by Yuki2718** [comment]
https://github.com/gorhill/uBlock/commit/869a653fdf2dd4e5b0151530bd58e1059d1318d2

**Reply by gorhill** [reply, marked as answer]
I will restore the ability in next dev build, I removed it and forgot to put it back with the new way.

**Reply by gorhill** [reply]
The JSON output can be often large, and the logger might not be optimal when trying to analyze these. I was trying to figure how I could make this easier but then I realized there is already a way to make this easier:

- Select "info" in filter expression picker
- Open export tool (the clipboard icon)
- Toggle to "List" mode
- Toggle to "Plain" mode

Now you can freely go through all output, select, copy/paste, etc.

**Reply by MasterKia** [reply]
> the logger might not be optimal when trying to analyze these but then I realized there is already a way to make this easier:

The console is still way better, I can't wrap my head around reading and analyzing these logs in the logger.
I might consider using an older uBO version just for having logs in the console.

**Reply by Yuki2718** [reply]
I think leaving from console is good, as we no more need to care about anti-devtools at least for this. It's just the current logger is not ideal.

**Reply by Yuki2718** [reply]
How about making the logger to have two tabs, the one is the old logger and the other is for these?

**Reply by Yuki2718** [reply]
It's also inconvenient I see all scriptlet logs when I don't want to see. I know I can exclude with `info`, but I think having two tabs is better. Not something too important though.

**Reply by Yuki2718** [reply]
And another problem is long lines are cut-off and we have to hover over them to see the whole logs.

**Reply by ghost** [reply]
First button on the toolbar. https://github.com/gorhill/uBlock/wiki/The-logger#expand

**Reply by Yuki2718** [reply]
Good, but tbh no need to expand all entries. I can't think of a better fix than separating scriptlet's output from the rest of the logs.

**Reply by gorhill** [reply]
> having two tabs

What do you call a "tab"? There are no tabs in the logger, just different views, accessible through toolbar icons.

**Reply by Yuki2718** [reply]
I mean create one as doesn't exist. Whatever way to switch between usual logs and scriptlet output with single click will be suffice.

**Reply by gorhill** [reply]
I am going to add a console-like pane in the logger plus other changes, but this will have to wait for next dev cycle otherwise this will delay a new stable release too far in the future.

**Comment by partingscientist** [comment]
Slightly off-topic, don't feel like opening another discussion for something minor.

@gorhill [this](https://github.com/gorhill/uBlock/blob/cd736b515ca75f21c6b88dd0db64d35f7aa8da67/src/js/scriptlets/scriptlet-loglevel-1.js#L29) is supposed to be `'setScriptletLogLevelToOne'`, no? To match [the receiver](https://github.com/gorhill/uBlock/blob/cd736b515ca75f21c6b88dd0db64d35f7aa8da67/assets/resources/scriptlets.js#L183)? Not sure what it breaks though.

**Reply by gorhill** [reply]
Thanks for spotting this, a typo indeed. This prevented from turning off verbose mode in the logger.

**Comment by stephenhawk8054** [comment]
@gorhill Sorry, do you know how to see the logs of `setTimeout` and `setInterval` in dev build?

**Reply by gorhill** [reply]
This will be in next dev build, if `prevent-setTimeout`, `prevent-setInterval` and `prevent-addEventListener` are used without arguments, they will simply output to logger without preventing.

**Reply by stephenhawk8054** [reply]
@ghajini Looks like you also asked about `no-fetch/xhr-if`?

**Reply by ghajini** [reply]
yeah i have

`1cloudfile.com/bqyX`
new updated filters, firefox
add 
`1cloudfile.com#@#+js(no-fetch-if, /adsbygoogle\.js|\.js/ method:HEAD)`
`1cloudfile.com##+js(no-fetch-if)`

i don't see scriptlet logging in logger & although not related to ublock origin, i see memory uprise in firefox due to antiadb code

**Reply by JobcenterTycoon** [reply]
Site issue. uBO will log when the crashing interval getting defused with `1cloudfile.com##+js(nosiif, display)`

**Reply by ghajini** [reply]
i have opera with ubo dev 1.55.b11 & above filters, console still shows scriptlet logging but no logging in ubo logger on  firefox ,ubo 1.55.1b23

**Reply by gorhill** [reply]
> Looks like you also asked about `no-fetch`/`xhr-if`

This will also be in next dev build, no argument = log all calls

**Reply by ghajini** [reply]
does logging breaks website?
`https://1cloudfile.com/bqyX`
firefox, ubo 1.55.1b24
```
1cloudfile.com#@#+js(no-fetch-if, /adsbygoogle\.js|\.js/ method:HEAD)
1cloudfile.com##+js(no-fetch-if)
```
**opera & ubo 1.55.1b11**
https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/6614ec7c-12c0-46c8-9e3d-3d4898ae2c96

**firefox & ubo 1.55.1b24**
https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/ec0ec161-9140-4e2e-96a5-984832c71058
https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/c1aa2ae3-e172-45cf-aac0-0bc42e834589

**Reply by gorhill** [reply]
I can reproduce fan spinning without logger, as per profiler this is caused by deep (infinite?) recursion in site's own `setInterval` handler:

image: https://github.com/uBlockOrigin/uBlock-issues/assets/585534/8c537d20-befd-4665-9e2b-6fff2679c5f4

**Reply by ghajini** [reply]
comment updated, please have a look

**Reply by gorhill** [reply]
This works for me:

    1cloudfile.com##+js(prevent-setInterval, !display)

**Reply by ghajini** [reply]
1.
i mean this should be like https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/6614ec7c-12c0-46c8-9e3d-3d4898ae2c96
when used
```
1cloudfile.com#@#+js(no-fetch-if, /adsbygoogle\.js|\.js/ method:HEAD)
1cloudfile.com##+js(no-fetch-if)
```

2. 
In fx & latest dev something is broken https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/ec0ec161-9140-4e2e-96a5-984832c71058

**Reply by gorhill** [reply]
> this should be like

Who knows? The site is throwing pile of code at the browser, and it's very possible the behavior is different somewhere in that pile of code, be it how `setInterval` is internally throttled by specific JS engine or whatever else. At least the the filter above narrow it down that that one pile of code passed to `setInterval`, but it's all obfuscated and since it does fix the issue I don't feel like digging deeper.

**Reply by gorhill** [reply]
> In fx & latest dev

You mean you compared latest Firefox dev build with stable Firefox built?

**Reply by gorhill** [reply]
Ok I see a difference between Firefox stable and dev build. I will investigate when I can.

**Reply by ghajini** [reply]
i used firefox stable only 
& filters
```
1cloudfile.com##+js(no-fetch-if, /adsbygoogle\.js|\.js/ method:HEAD)
1cloudfile.com##+js(no-fetch-if)
1cloudfile.com##+js(nosiif, !display)
```

1. ubo stable 1.55 =
https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/b4177578-f6a1-4d05-a5aa-353bd3df9830

2. ubo dev 1.55.1b24=

https://github.com/uBlockOrigin/uBlock-issues/assets/20338483/0d637619-6a3e-48a6-9c75-beb2e2548fc7

**expected is i should see `Please Disable adblock ...` when only `1cloudfile.com##+js(no-fetch-if)` is used** but recent ubo dev is breaking

**Reply by gorhill** [reply]
> recent ubo dev is breaking

We shouldn't say "breaking" just because the behavior is different. The site appears to be trapping calls to `console.log`, and since the new dev build doesn't use `console.log` for the involved scriptlets, it's possible this is what is causing the different behavior.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/585534/519b74f5-f54c-4739-92fe-c477cbeb33d9)

**Reply by ghajini** [reply]
the message about antiadb is hidden when used logging scriptlet `no-fetch-if` , i think can be fixed

**Comment by dimisa-RUAdList** [comment]
In version 1.55.1b25 I observe unusual behavior when adding: `##+js(no-xhr-if)`. In addition to outputting the log in the log, the scriptlet begins to intercept calls.

Example: `https://anidubonline.com/anime/full/11635-sozdannyj-v-bezdne-tv-2-made-in-abyss-retsujitsu-no-ougonkyou-01-iz-12.html` 

Add: `anidubonline.com##+js(no-xhr-if)`, select an episode and press play.

**Reply by gorhill** [reply]
https://github.com/gorhill/uBlock/commit/1db54c47e134c1f093fa4e7aeed7caf902e286a3

**Comment by dimisa-RUAdList** [comment]
@gorhill 
`example.com##+js(aeld, _, , log, 3)` - will it be implemented in upcoming releases?

**Reply by gorhill** [reply]
What are you trying to accomplish? Use `example.com##+js(aeld)` to log all calls to `addEventListener`.

**Reply by dimisa-RUAdList** [reply]
The thing is that when using 1.55.1b26 I do not see calls in the log, so I thought that this had not yet been implemented.

`https://overclockers.ru/`

`overclockers.ru##+js(aeld, _, , log, 3)`

**Reply by gorhill** [reply]
Use `overclockers.ru##+js(aeld)`, the old way to trigger on logging does not exist anymore, so when you have `_` as second parameter, the scriptlet is no longer in log-all mode.

**Reply by dimisa-RUAdList** [reply]
Great, it works. Thank you!

**Reply by dimisa-RUAdList** [reply]
I test the rules on different systems and browsers. Therefore, I did not immediately understand what the problem was. Now that I knew the correct rules for displaying the log, I was able to find out that the problem occurs only in Google Chrome 109. In 9 cases out of 10, the log does not display calls. Sometimes this can be fixed by double-clicking on the icon `</>` in the logger. But this doesn't always help.

**Reply by gorhill** [reply]
I am really not equipped to test old versions of Chromium.

**Reply by Yuki2718** [reply]
Will I see the logs for `japscan.lol##+js(aeld)` in the logger on 1.55.1b26? I don't see anything.

**Reply by gorhill** [reply]
I can reproduce issue on `japscan.lol`, so maybe the one on Google Chrome 109 is related.

**Reply by Yuki2718** [reply]
Also how to log nowoif? With `japscan.lol##+js(nowoif, , , log)` I only see `[www.japscan.lol][no-window-open-if ⁝  ⁝  ⁝ log] Prevented` but not the url it prevented.

**Reply by Yuki2718** [reply]
Hmmm, `japscan.lol##+js(nowoif, , , log)` doesn't log on 1.55 too, nevetheless the rule works.

**Reply by gorhill** [reply]
Do not use `log` anymore in any scriptlet, this was useful only when outputing to console. Now if logger is opened, it will output whatever there is to output. If `japscan.lol##+js(nowoif)` doesn't work, then I need to fix it -- I just need feedback about what trace information you want to see in logger. So I will add arguments information for `nowoif`. Also, if verbose mode in enabled, it will also output not-prevented calls to `window.open`.

**Reply by Yuki2718** [reply]
Okay, `japscan.lol##+js(nowoif)` logs this

![logs](https://github.com/uBlockOrigin/uBlock-issues/assets/58900598/d5fd416d-c4d0-4a19-a951-fd07c1193418)

I want to know what `window.open` arguments, in particular url, it prevented. But the situation is the same for 1.55.0 or AG, maybe some problem on Chrome or the site.

**Reply by gorhill** [reply]
II modified the scriptlet to return more information -- in next dev build.

From now on I won't be able to do anything today and in the coming days, so hopefully the next devbuild will be good enough.

**Reply by dimisa-RUAdList** [reply]
Yes, in version 1.55.1b27 in Google Chrome 109 the problem (aeld) is solved.

**Comment by stephenhawk8054** [comment]
@gorhill Sorry, do you know how to log `aost`?

**Reply by ghajini** [reply]
`example.com##+js(aost, property)`
https://github.com/uBlockOrigin/uBlock-discussions/discussions/841#discussioncomment-7777610

**Reply by stephenhawk8054** [reply]
Thanks. Confirm it works.

**Comment by gorhill** [comment]
I just discovered an issue with the sending of log information to the logger. The issue can be triggered when adding the filter `arstechnica.com##+js(json-prune)` and visiting `https://arstechnica.com/science/2024/03/redwoods-are-growing-almost-as-fast-in-the-uk-as-their-californian-cousins/` (could be any page probably).

Sometimes there are so many logged entries sent to the logger, that [this line](https://github.com/gorhill/uBlock/blob/1.56.0/platform/common/vapi-client.js#L206) will become true, which will cause all the injected scriptlets to essentially abort because the communication channel is closed for all content scripts. I found the issue when trying to launch the DOM inspector, which was failing. So for now I will raise the limit to 1000 while trying to figure out if a better fix exists.
